Allow registering the same subtype multiple times

jkosh44 opened the pull request:
Before this commit NamedType hashes only on it's class and not on the
name. This only allowed you to register a class once using
ObjectMapper.registerSubtypes(NamedType... types). This commit now
uses name field to hash NamedTypes.

This successfully allows you to deserialize objects of the same type
but different name.

Serializing objects of the same type but different name (for example
fields of a POJO) still has some issues.
1. SerializerProvider caches Serializers based on class and doesn't
take name into account.
2. TypeIdResolver has no method to resolve an id that takes name into
account. Therefore when resolving an id we only have the value and
type.
3. TypeNameIdResolver stores type and id information as a
Map<String, String> that maps type to id, ignoring name

Fixes #2515

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/jsontype/NamedType.java:
```diff
@@ -14,10 +14,10 @@
     public NamedType(Class<?> c, String name) {
         _class = c;
-        _hashCode = c.getName().hashCode();
+        _hashCode = (c.getName() + (name != null ? name : "")).hashCode();
```
Will also need to change `equals()` to compare name now (to be compatible with semantic definition of `hashCode()`), I think.

cowtowncoder commented on src/test/java/com/fasterxml/jackson/databind/jsontype/TestSubtypes.java:
```diff
@@ -216,8 +239,12 @@ public void testSerialization() throws Exception
         mapper.registerSubtypes(new NamedType(SubB.class, "typeB"));
         assertEquals("{\"@type\":\"typeB\",\"b\":1}", mapper.writeValueAsString(bean));
 
+        // the first registered type name is used for serialization
```
It'd probably be better to leave existing test as is, but also add a new extended one.

jkosh44 commented:
I addressed the comments and re-pushed.

Also as mentioned in the commit message this doesn't really solve the full use case and the Serialization part still needs some work. I just think that will be much more in depth, and maybe should get it's own PR.

cowtowncoder commented:
@jkosh44 Thank you again for contributing this -- it will be in 2.11.0.

I think it is good to make incremental progress: so, deserialization works as expected now which is a solid improvement. Serialization side may be improved with more work if that seems important; and we can also consider adding a feature to allow alternatively failing on case where it is not possible to decide between conflicting settings.
